subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/JvmSoftwareComponentInternal.java

```diff
@@ -0,0 +1,165 @@
+ * is the lack of multi-target functionality. This interface currently exposes the
+ * {@link #getJar()}, {@link #getSources()}, {@link #getImplementation()}, etc. methods, which
+ * imply that this component only supports a single compilation target. A multi-target component
+ * would not expose these methods at the top-level, but would instead encapsulate them within
+ * some kind of {@code targets} or {@code pipelines} container.
+ * <p>
+ * Before publicizing this component, we also need to consider how component extensibility works.
+ * For example, for the java-library plugin we have the additional {@code api} and {@code compileOnlyApi}
+ * scopes. Do we expect that plugin to write a new component interface and extend this in order to add the
+ * proper getter methods? What about the concrete component class which implements that new interface? Does
+ * it extend the default implementation class? Is there a way we can avoid Java inheritance?
+ */
+public interface JvmSoftwareComponentInternal extends SoftwareComponent {
+
+    /**
+     * Configures this component to publish a javadoc jar alongside the primary artifacts. As a result,
+     * this method also configures the necessary configurations and tasks required to produce
+     * the javadoc artifact.
+     */
+    void withJavadocJar();
+
+    /**
+     * Configures this component to publish a sources jar alongside the primary artifacts. As a result,
+     * this method also configures the necessary configurations and tasks required to produce
+     * the sources artifact.
+     */
+    void withSourcesJar();
+
+    /**
+     * Get the {@link Jar} task which assembles the resources and compilation outputs into
+     * a single artifact.
+     *
+     * @return A provider which supplies the component's {@link Jar} task.
+     */
+    TaskProvider<Jar> getJar();
+
+    /**
+     * Get the {@link JavaCompile} task which compiles the Java source files into classes.
+     *
+     * @return A provider which supplies the component's {@link JavaCompile} task.
+     */
+    TaskProvider<JavaCompile> getCompileJava();
+
+    // TODO: This could use a better name. These "outputs" can be confused with a
+    // component's "real" outputs, which are its variants.
+    /**
+     * Get the resources and compilation outputs for this component which are used as
+     * inputs for the {@link #getJar() jar} task.
+     *
+     * @return The source set outputs.
+     */
+    SourceSetOutput getOutput();
+
+    /**
+     * Get this component's backing source set.
+     * <p>
+     * {@link SourceSet#getOutput()} and the classpath-returning methods on the returned
+     * source set should ideally be avoided in favor of the similarly-named methods on
+     * this component. The concept of source sets having a single set of outputs is only
+     * relevant for single-target components.
+     *
+     * @return This component's source set.
+     */
+    SourceSet getSources();
```

Comment:
🤔 This sounds plural to me.  I think `getSourceSet` would be better.

Reply:
I was just following the lead of `JvmTestSuite`: 
https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmTestSuite.java#L47-L54

If we rename this we should rename the one there as well.